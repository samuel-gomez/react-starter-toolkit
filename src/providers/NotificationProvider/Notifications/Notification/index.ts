export { default } from './Notification.container';
export { EType } from './Notification';
export type { TNotification } from './Notification';
