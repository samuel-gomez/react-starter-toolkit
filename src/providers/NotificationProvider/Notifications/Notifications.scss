@import '@axa-fr/react-toolkit-core/dist/assets/scss/core.scss';

$typesLite: (
  success $color-alertSucces-leftBgColor $color-alertSucces-bgColor $color-alertSucces-colorBorder $color-alertSucces-textColor,
  info $color-alertInfo-leftBgColor $color-alertInfo-bgColor $color-alertInfo-colorBorder $color-alertInfo-textColor,
  danger $color-alertDanger-leftBgColor $color-alertDanger-bgColor $color-alertDanger-colorBorder $color-alertDanger-textColor,
  error $color-alertError-leftBgColor $color-alertError-bgColor $color-alertError-colorBorder $color-alertError-textColor
);

.af {
  &-notifications {
    position: fixed;
    width: 400px;
    background: rgba(0, 0, 0, 0);
    padding: 1rem;
    top: 0;
    right: 0;
    z-index: 13;
    &--open {
      animation: 0.5s cubic-bezier(0.08, 0.82, 0.17, 1) forwards slideToLeft;
    }
  }
  &-alert--notification {
    overflow: hidden;
    max-height: 300px;
    box-shadow: 2px 2px 2px 0 rgba(0, 0, 0, 0.2);
    transition: opacity 0.3s ease-out, max-height 0.3s ease-out 0.3s, margin 0.3s ease-out 0.3s;
    opacity: 1;
    &.af-alert--hide {
      opacity: 0;
      max-height: 0;
      margin: 0;
    }
    .af-alert {
      &__title-icon {
        background-color: inherit;
        min-width: inherit;
        padding: 0 0 0 1rem;
        &:after {
          display: none;
        }
      }
      &__title-text {
        padding: 0.9375rem 1.5rem 0.9375rem 1rem;
        font-size: 1rem;
        color: $color-texte;
      }
      &__content {
        line-height: 0.9rem;
        font-size: 0.9rem;
        .af-alert__content__left {
          background: inherit;
          width: 1rem;
          min-width: 1rem;
          margin-left: 1rem;
        }
        &__right {
          color: $color-texte;
        }
      }
    }
  }
  &-alert {
    @each $typeLite in $typesLite {
      &--#{nth($typeLite, 1)}.af-alert--notification {
        border: 0;
        border-left: 2px solid nth($typeLite, 2);
        .af-alert {
          &__title-icon {
            .glyphicon {
              color: nth($typeLite, 2);
            }
          }
        }
      }
    }
  }
}

@include media-breakpoint-down(sm) {
  .af {
    &-notifications {
      width: 100%;
    }
  }
}
