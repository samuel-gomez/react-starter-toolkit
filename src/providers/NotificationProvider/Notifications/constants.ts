export const DEFAULT_CLASS_NOTIFICATION = 'af-notifications';
export const MODIFIER_OPEN = 'open';
export const MODIFIER_HIDE = 'hide';
export const DELAY_HIDE = 500;

export const ALERT_ICON = {
  success: 'ok',
  error: 'minus-sign',
  danger: 'alert',
};
