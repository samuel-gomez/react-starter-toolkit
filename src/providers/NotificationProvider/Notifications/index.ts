export { default } from './Notifications.container';
export type { TNotifications } from './Notifications';
export type { TNotification, EType } from './Notification';
