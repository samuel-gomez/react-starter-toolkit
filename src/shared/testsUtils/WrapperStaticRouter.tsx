/* eslint-disable react/display-name */
import { ReactNode } from 'react';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';

const WrapperStaticRouter =
  (route = '/') =>
  ({ children }: { children: ReactNode }) =>
    <MemoryRouterProvider url={route}>{children}</MemoryRouterProvider>;

export default WrapperStaticRouter;
