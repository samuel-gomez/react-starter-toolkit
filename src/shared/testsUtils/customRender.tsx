/* eslint-disable react/display-name */
import { render, RenderOptions } from '@testing-library/react';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';

import FetchProvider from 'providers/FetchProvider';
import UserProvider from 'providers/UserProvider';
import { ReactElement, ReactNode } from 'react';
import { MOCK_API_URL } from '.';

type TMockProvider = {
  [x: string]: Record<string, unknown | number | string> | string;
};

const MockProviders =
  ({ role = '', ...testMock }: TMockProvider) =>
  ({ children }: { children: ReactNode }) => {
    const useOidcAccessTokenFn = jest.fn().mockReturnValue({ accessToken: 'accessToken' });
    const useOidcUserFn = jest.fn().mockReturnValueOnce({ oidcUser: { member_of: [`CN=${role}`], name: 'Samuel Gomez' } });
    return (
      <UserProvider useOidcUserFn={useOidcUserFn}>
        <FetchProvider
          apiUrl={MOCK_API_URL}
          fetchConfig={{
            headers: { testMock: JSON.stringify(testMock) },
          }}
          useOidcAccessTokenFn={useOidcAccessTokenFn}
        >
          <MemoryRouterProvider url="">{children}</MemoryRouterProvider>
        </FetchProvider>
      </UserProvider>
    );
  };

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>, testMock: TMockProvider = { role: '' }) =>
  render(ui, { wrapper: MockProviders(testMock), ...options });

// re-export everything
export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

// override render method
export { customRender as render };
