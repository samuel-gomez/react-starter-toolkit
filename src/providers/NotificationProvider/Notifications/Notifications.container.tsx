import { ClassManager } from '@axa-fr/react-toolkit-core';
import Notifications, { TNotifications } from './Notifications';
import { DEFAULT_CLASS_NOTIFICATION, MODIFIER_OPEN } from './constants';

type TNotificationsContainer = Omit<TNotifications, 'notifications'> & {
  notifications?: TNotifications['notifications'];
  getComponentClassName?: typeof ClassManager.getComponentClassName;
  NotificationsCmpt?: typeof Notifications;
};

const NotificationsContainer = ({
  className = DEFAULT_CLASS_NOTIFICATION,
  notifications = [],
  getComponentClassName = ClassManager.getComponentClassName,
  NotificationsCmpt = Notifications,
  ...rest
}: TNotificationsContainer) => {
  const classModifier = notifications.length > 0 ? MODIFIER_OPEN : '';
  const classComponent = getComponentClassName(className, classModifier, DEFAULT_CLASS_NOTIFICATION);
  return <NotificationsCmpt {...rest} notifications={notifications} className={classComponent} />;
};

export default NotificationsContainer;
