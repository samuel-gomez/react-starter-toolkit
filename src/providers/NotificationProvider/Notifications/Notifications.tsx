import './Notifications.scss';
import { DEFAULT_CLASS_NOTIFICATION } from './constants';
import Notification, { TNotification, EType } from './Notification';
import type { TonDeleteNotification } from '..';

export type TNotifications = {
  className?: string;
  notifications: TNotification[];
  deleteNotification: TonDeleteNotification;
};

const Notifications = ({ notifications, deleteNotification, className = DEFAULT_CLASS_NOTIFICATION }: TNotifications) => (
  <aside className={className}>
    {notifications.map(({ id, label, type = EType.error, classModifier, ...rest }) => (
      <Notification {...rest} key={id} id={id} label={label} type={type} classModifier={classModifier} onClose={() => deleteNotification?.(id)} />
    ))}
  </aside>
);

export default Notifications;
