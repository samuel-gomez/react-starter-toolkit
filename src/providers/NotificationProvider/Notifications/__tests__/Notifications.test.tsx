import { render, fireEvent, screen } from '@testing-library/react';
import { emptyFunction } from 'shared/testsUtils';
import Notifications from '../Notifications';

const defaultProps = {
  notifications: [],
  deleteNotification: emptyFunction,
};

describe('Notifications', () => {
  it('Render <Notifications/>', () => {
    const { asFragment } = render(<Notifications {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should contain Notification when array notifications is NOT empty', () => {
    const notifications = [{ id: 'id', label: 'titleerror', classModifier: '', onClose: emptyFunction }];
    const { asFragment, getByText } = render(<Notifications {...defaultProps} notifications={notifications} />);
    expect(getByText('titleerror')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should called onDeleteNotification with id When click on close button', () => {
    const deleteNotificationMock = jest.fn();
    const notifications = [{ id: 'id', label: 'title' }];
    render(<Notifications {...defaultProps} notifications={notifications} deleteNotification={deleteNotificationMock} />);
    fireEvent.click(screen.getByRole('button'));
    expect(deleteNotificationMock).toBeCalledWith('id');
  });
});
