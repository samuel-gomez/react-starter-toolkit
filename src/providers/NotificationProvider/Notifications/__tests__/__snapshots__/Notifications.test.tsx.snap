// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications Render <Notifications/> 1`] = `
<DocumentFragment>
  <aside
    class="af-notifications"
  />
</DocumentFragment>
`;

exports[`Notifications Should contain Notification when array notifications is NOT empty 1`] = `
<DocumentFragment>
  <aside
    class="af-notifications"
  >
    <div
      class="af-alert af-alert--notification af-alert--error"
      role="alert"
    >
      <div
        class="af-alert__title"
      >
        <div
          class="af-alert__title-icon"
        >
          <i
            class="glyphicon glyphicon-minus-sign"
          />
        </div>
        <div
          class="af-alert__title-text"
        >
          titleerror
        </div>
        <button
          class="af-alert__title-icon-close"
          type="button"
        >
          <i
            class="glyphicon glyphicon-close"
          />
        </button>
      </div>
    </div>
  </aside>
</DocumentFragment>
`;
